docs/architecture/011-api-key-access-control.md

```diff
@@ -0,0 +1,87 @@
+# 011: API Key Access Control
+
+Authors: @opqdonut
+
+## Background
+
+REMS access control has always been user-based. For example all
+application commands must have an actor, and the code checks whether
+this actor is allowed to perform this command. API authentication has
+been done with two headers: `x-rems-api-key` which must be a valid api
+key, and `x-rems-user-id`, which specifies which user is being
+impersonated.
+
+Some users have had a need to limit the access available with API
+keys. In response, we added a list of allowed roles to each API key.
+The roles were a whitelist, with which the roles of the impersonated
+user were filtered.
+
+This approach works fine for so-called _explicit roles_ that are
+granted using the `role` table in the database. However, we also have
+_implicit roles_ that are granted per application, based on the
+application state. For example, a user has the `:reviewer` role for an
+application if they have been invited to review the application. When
+the API key roles were implemented, they were (accidentally?) only
+implemented for explicit roles. Extending the implementation for
+implicit roles would require more code.
+
+As a concrete example, an API key with only the role `:logged-in`
+would still be able to impersonate the handler of an application and
+perform handler commands like approving the applicaiton.
```

Comment:
> application.

Reply:
How about add here the mention of the background why the previous system was implemented, e.g. the requirements from NKR which will be elaborated in the use cases? That's why we also consider this change now.

Reply:
Typo fixed, paragraph added.